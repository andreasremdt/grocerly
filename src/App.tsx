import { useContext } from "react";

import Header from "./components/Header";
import List from "./components/List";
import Form from "./components/Form";
import Settings from "./components/Settings";
import ListOverview from "./components/ListOverview";
import { GroceryContext } from "./GroceryContext";

function App() {
  const { activeList, isSettingsVisible } = useContext(GroceryContext);

  return (
    <>
      <Header />

      {isSettingsVisible ? (
        <Settings />
      ) : (
        <>
          <Form />
          {activeList ? <List /> : <ListOverview />}
        </>
      )}
    </>
  );
}

export default App;
