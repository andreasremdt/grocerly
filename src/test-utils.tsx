import { ReactNode, Dispatch } from "react";
import { render, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter, Routes, Route } from "react-router-dom";

import { GroceryContext } from "./GroceryContext";
import { GroceryState, GroceryActions } from "./types";

const DEFAULT_CONTEXT_STATE: GroceryState = {
  groceries: [],
  lists: [],
  editing: null,
  isFormVisible: false,
  isNewListDialogVisible: false,
  settings: {
    language: "en",
    sortByChecked: true,
  },
};

type GroceryContextState = GroceryState & {
  dispatch: Dispatch<GroceryActions>;
};

function renderWithContext(
  children: ReactNode,
  props?: Partial<GroceryContextState>,
  options?: { withRoutes?: boolean; initialEntries?: string[]; clean?: boolean }
) {
  const state: GroceryContextState = { ...DEFAULT_CONTEXT_STATE, dispatch: jest.fn(), ...props };

  if (options?.clean) {
    cleanup();
  }

  return render(
    <MemoryRouter initialEntries={options?.initialEntries || ["/"]}>
      <GroceryContext.Provider value={state}>{children}</GroceryContext.Provider>
      {options?.withRoutes && (
        <Routes>
          <Route path="/" element={<h1>Home page</h1>} />
          <Route path="/settings" element={<h1>Settings page</h1>} />
          <Route path="/list/:listId" element={<h1>List page</h1>} />
        </Routes>
      )}
    </MemoryRouter>
  );
}

export * from "@testing-library/react";
export { userEvent };
export { renderWithContext as render };
