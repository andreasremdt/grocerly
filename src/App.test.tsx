import { fireEvent, render } from "@testing-library/react";
import { act } from "react-dom/test-utils";

import App from "./App";

window.navigator.vibrate = jest.fn();

beforeEach(() => {
  localStorage.clear();
});

test("the app structure is rendered correctly", () => {
  const { container } = render(<App />);

  expect(container).toMatchSnapshot();
});

test("toggles the settings page and switches colors", () => {
  const { getByTitle, getByText, queryByText, getByLabelText, rerender } = render(<App />);

  expect(queryByText(/theme color/i)).not.toBeInTheDocument();
  fireEvent.click(getByTitle(/toggle settings/i));
  expect(getByText(/theme color/i)).toBeInTheDocument();
  fireEvent.click(getByLabelText(/orange/i));
  expect(document.body.style.getPropertyValue("--primary-color")).toEqual("var(--orange)");
  expect(localStorage.getItem("color")).toEqual("orange");
  fireEvent.click(getByLabelText(/blue/i));
  expect(document.body.style.getPropertyValue("--primary-color")).toEqual("var(--blue)");
  expect(localStorage.getItem("color")).toEqual("blue");

  rerender(<App />);

  expect(document.body.style.getPropertyValue("--primary-color")).toEqual("var(--blue)");
  expect((getByLabelText(/blue/i) as HTMLInputElement).checked).toEqual(true);
});

test("switches the language", () => {
  const { getByTestId, getByTitle, getByText, queryByText } = render(<App />);

  expect(localStorage.getItem("language")).toEqual("en");
  fireEvent.click(getByTitle(/toggle settings/i));
  fireEvent.change(getByTestId("language-select"), { target: { value: "de" } });
  expect(queryByText(/nothing here, yet/i)).not.toBeInTheDocument();
  expect(getByText(/diese liste ist noch leer./i)).toBeInTheDocument();
  expect(localStorage.getItem("language")).toEqual("de");
});

test("toggles the form", () => {
  const { getByRole, getByTitle, queryByRole, getByPlaceholderText } = render(<App />);

  expect(getByRole("textbox")).toBeInTheDocument();
  expect(getByPlaceholderText(/eggs/i)).toHaveFocus();
  fireEvent.click(getByTitle(/toggle form/i));
  expect(queryByRole("textbox")).not.toBeInTheDocument();
  fireEvent.click(getByTitle(/toggle form/i));
  expect(getByRole("textbox")).toBeInTheDocument();
  expect(getByPlaceholderText(/eggs/i)).toHaveFocus();
});

test("items are persisted in local storage", () => {
  Element.prototype.scrollTo = jest.fn();

  const { getByPlaceholderText, getByLabelText, getByText, getAllByTitle } = render(<App />);

  fireEvent.change(getByPlaceholderText(/eggs/i), { target: { value: "tomatoes" } });
  fireEvent.change(getByPlaceholderText(/qt/i), { target: { value: "3" } });
  fireEvent.click(getByText(/^add$/i));

  expect(JSON.parse(localStorage.getItem("data")!)[0]).toMatchObject(
    expect.objectContaining({
      amount: "3",
      name: "tomatoes",
    })
  );

  fireEvent.change(getByPlaceholderText(/eggs/i), { target: { value: "water" } });
  fireEvent.change(getByPlaceholderText(/qt/i), { target: { value: "1" } });
  fireEvent.click(getByLabelText(/^l$/i));
  fireEvent.click(getByText(/^add$/i));

  expect(JSON.parse(localStorage.getItem("data")!).length).toEqual(2);
  fireEvent.click(getAllByTitle(/delete item/i)[0]);
  expect(JSON.parse(localStorage.getItem("data")!).length).toEqual(1);
});

test("adds, updates, and removes items", () => {
  Element.prototype.scrollTo = jest.fn();
  jest.useFakeTimers("modern");

  const { getByPlaceholderText, getByLabelText, getByText, queryByText, getByTitle } = render(
    <App />
  );

  expect(getByText(/nothing here, yet/i)).toBeInTheDocument();

  fireEvent.change(getByPlaceholderText(/eggs/i), { target: { value: "milk" } });
  fireEvent.change(getByPlaceholderText(/qt/i), { target: { value: "100" } });
  fireEvent.click(getByLabelText(/ml/i));
  fireEvent.click(getByText(/^add$/i));

  expect(getByText(/100ml/i)).toBeInTheDocument();
  expect(getByText(/milk/i)).toBeInTheDocument();
  expect(getByPlaceholderText(/eggs/i)).toHaveFocus();
  expect((getByPlaceholderText(/eggs/i) as HTMLInputElement).value).toEqual("");
  expect((getByPlaceholderText(/qt/i) as HTMLInputElement).value).toEqual("");
  expect((getByLabelText(/ml/i) as HTMLInputElement).checked).toBeFalsy();
  expect(Element.prototype.scrollTo).toHaveBeenCalledTimes(1);

  act(() => {
    fireEvent.click(getByText(/milk/i).parentElement!, {
      detail: 1,
    });
    jest.runAllTimers();
  });

  expect(getByText(/milk/i).parentElement!).toHaveStyle("text-decoration: line-through");

  act(() => {
    fireEvent.click(getByText(/milk/i).parentElement!, {
      detail: 1,
    });
    jest.runAllTimers();
  });

  expect(getByText(/milk/i).parentElement!).toHaveStyle("text-decoration: none");
  fireEvent.dblClick(getByText(/milk/i).parentElement!);
  expect((getByPlaceholderText(/eggs/i) as HTMLInputElement).value).toEqual("milk");
  expect((getByPlaceholderText(/qt/i) as HTMLInputElement).value).toEqual("100");
  expect((getByLabelText(/ml/i) as HTMLInputElement).checked).toBeTruthy();
  fireEvent.change(getByPlaceholderText(/eggs/i), { target: { value: "bread" } });
  fireEvent.click(getByText(/^add$/i));

  expect(getByText(/100ml/i)).toBeInTheDocument();
  expect(queryByText(/milk/i)).not.toBeInTheDocument();
  expect(getByText(/bread/i)).toBeInTheDocument();

  fireEvent.click(getByTitle(/delete item/i));
  expect(getByText(/nothing here, yet/i)).toBeInTheDocument();
  expect(Element.prototype.scrollTo).toHaveBeenCalledTimes(1);
});
