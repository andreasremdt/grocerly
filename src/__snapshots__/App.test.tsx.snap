// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the app structure is rendered correctly 1`] = `
<div>
  <header
    class="wrapper"
  >
    <h1
      class="title"
    >
      Grocerly
    </h1>
    <button
      class="button active"
      title="Toggle form"
      type="button"
    >
      <svg
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
      >
        <polyline
          points="18 15 12 9 6 15"
        />
      </svg>
    </button>
    <button
      class="button"
      title="Delete all items"
      type="button"
    >
      <svg
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
      >
        <polyline
          points="3 6 5 6 21 6"
        />
        <path
          d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
        />
        <line
          x1="10"
          x2="10"
          y1="11"
          y2="17"
        />
        <line
          x1="14"
          x2="14"
          y1="11"
          y2="17"
        />
      </svg>
    </button>
    <button
      class="button "
      title="Toggle settings"
      type="button"
    >
      <svg
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
      >
        <circle
          cx="12"
          cy="12"
          r="3"
        />
        <path
          d="M19.4 15a1.65 1.65 0 0 0 .33 1.82l.06.06a2 2 0 0 1 0 2.83 2 2 0 0 1-2.83 0l-.06-.06a1.65 1.65 0 0 0-1.82-.33 1.65 1.65 0 0 0-1 1.51V21a2 2 0 0 1-2 2 2 2 0 0 1-2-2v-.09A1.65 1.65 0 0 0 9 19.4a1.65 1.65 0 0 0-1.82.33l-.06.06a2 2 0 0 1-2.83 0 2 2 0 0 1 0-2.83l.06-.06a1.65 1.65 0 0 0 .33-1.82 1.65 1.65 0 0 0-1.51-1H3a2 2 0 0 1-2-2 2 2 0 0 1 2-2h.09A1.65 1.65 0 0 0 4.6 9a1.65 1.65 0 0 0-.33-1.82l-.06-.06a2 2 0 0 1 0-2.83 2 2 0 0 1 2.83 0l.06.06a1.65 1.65 0 0 0 1.82.33H9a1.65 1.65 0 0 0 1-1.51V3a2 2 0 0 1 2-2 2 2 0 0 1 2 2v.09a1.65 1.65 0 0 0 1 1.51 1.65 1.65 0 0 0 1.82-.33l.06-.06a2 2 0 0 1 2.83 0 2 2 0 0 1 0 2.83l-.06.06a1.65 1.65 0 0 0-.33 1.82V9a1.65 1.65 0 0 0 1.51 1H21a2 2 0 0 1 2 2 2 2 0 0 1-2 2h-.09a1.65 1.65 0 0 0-1.51 1z"
        />
      </svg>
    </button>
  </header>
  <form
    autocapitalize="off"
    class="wrapper"
    novalidate=""
  >
    <div
      class="amount-controls"
    >
      <button
        class="amount-button"
        disabled=""
        type="button"
      >
        -
      </button>
      <input
        class="input amount-input"
        placeholder="Qt."
        type="number"
        value=""
      />
      <button
        class="amount-button"
        type="button"
      >
        +
      </button>
    </div>
    <input
      class="input name-input"
      placeholder="e.g. eggs, milk, spinach"
      type="text"
      value=""
    />
    <div
      class="units"
    >
      <label
        class="label"
      >
        <input
          class="input"
          name="unit"
          type="radio"
          value="mg"
        />
        <span
          class="display"
        >
          mg
        </span>
      </label>
      <label
        class="label"
      >
        <input
          class="input"
          name="unit"
          type="radio"
          value="g"
        />
        <span
          class="display"
        >
          g
        </span>
      </label>
      <label
        class="label"
      >
        <input
          class="input"
          name="unit"
          type="radio"
          value="kg"
        />
        <span
          class="display"
        >
          kg
        </span>
      </label>
      <label
        class="label"
      >
        <input
          class="input"
          name="unit"
          type="radio"
          value="ml"
        />
        <span
          class="display"
        >
          ml
        </span>
      </label>
      <label
        class="label"
      >
        <input
          class="input"
          name="unit"
          type="radio"
          value="l"
        />
        <span
          class="display"
        >
          l
        </span>
      </label>
    </div>
    <button
      class="button"
      type="submit"
    >
      Add
    </button>
  </form>
  <main
    class="wrapper"
  >
    <div
      class="empty"
    >
      <svg>
        <use
          xlink:href="/shopping-cart.svg#img"
        />
      </svg>
      <p>
        <b>
          Nothing here, yet.
        </b>
        <br />
        Start adding items to your shopping list below.
      </p>
    </div>
  </main>
</div>
`;
